src/block_service/test/hdfs_service_test.cpp

```diff
@@ -0,0 +1,327 @@
+#include <dsn/dist/block_service.h>
+#include <dsn/utility/filesystem.h>
+#include <dsn/utility/flags.h>
+#include <dsn/utility/rand.h>
+#include <fstream>
+#include <gtest/gtest.h>
+#include <memory>
+
+#include "block_service/hdfs/hdfs_service.h"
+
+using namespace dsn;
+using namespace dsn::dist::block_service;
+
+static std::string example_name_node = "<hdfs_name_none>";
+static std::string example_backup_path = "<hdfs_path>";
+// Please modify following paras in 'config-test.ini' to enable hdfs_service_test,
+// or hdfs_service_test will be skipped and return true.
+DSN_DEFINE_string("hdfs_test", test_name_node, "<hdfs_name_none>", "hdfs name node");
+DSN_DEFINE_string("hdfs_test",
+                  test_backup_path,
+                  "<hdfs_path>",
+                  "path for uploading and downloading test files");
+
+DSN_DEFINE_uint32("hdfs_test", num_test_file_lines, 4096, "number of lines in test file");
+DSN_DEFINE_uint32("hdfs_test",
+                  num_total_files_for_hdfs_concurrent_test,
+                  64,
+                  "number of total files for hdfs concurrent test");
+
+DEFINE_TASK_CODE(LPC_TEST_HDFS, TASK_PRIORITY_HIGH, dsn::THREAD_POOL_DEFAULT)
+
+class HDFSClientTest : public testing::Test
+{
+protected:
+    virtual void SetUp() override;
+    virtual void TearDown() override;
+    void generate_test_file(const char *filename);
+    std::string name_node;
+    std::string backup_path;
+};
+
+void HDFSClientTest::SetUp()
+{
+    name_node = FLAGS_test_name_node;
+    backup_path = FLAGS_test_backup_path;
+}
+
+void HDFSClientTest::TearDown() {}
+
+void HDFSClientTest::generate_test_file(const char *filename)
+{
+    // generate a local test file.
+    int lines = FLAGS_num_test_file_lines;
+    FILE *fp = fopen(filename, "wb");
+    for (int i = 0; i < lines; ++i) {
+        fprintf(fp, "%04d_this_is_a_simple_test_file\n", i);
+    }
+    fclose(fp);
+}
+
+TEST_F(HDFSClientTest, test_basic_operation)
+{
+    if (name_node == example_name_node || backup_path == example_backup_path) {
+        return;
+    }
+
+    std::vector<std::string> args = {name_node, backup_path};
+    std::shared_ptr<hdfs_service> s = std::make_shared<hdfs_service>();
+    ASSERT_EQ(dsn::ERR_OK, s->initialize(args));
+
+    std::string local_test_file = "test_file";
+    std::string remote_test_file = "hdfs_client_test/test_file";
+    int64_t test_file_size = 0;
+
+    generate_test_file(local_test_file.c_str());
+    dsn::utils::filesystem::file_size(local_test_file, test_file_size);
+
+    create_file_response cf_resp;
+    ls_response l_resp;
```

Comment:
I think it's better to move these variables closer to the place where you use it

Reply:
done.